src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver to be checked.
+
+        Raises
+        ------
+        ValueError
+            If the provided solver name is not in the list of allowed optimizers.
+        """
+        if solver_name not in self.allowed_optimizers:
+            raise ValueError(
+                f"Solver `{solver_name}` not allowed for "
+                f"{self.__class__} regularization. "
+                f"Allowed solvers are {self.allowed_optimizers}."
+            )
+
+    @staticmethod
+    def _check_solver_kwargs(solver_name, solver_kwargs):
+        """
+        Check if provided solver keyword arguments are valid.
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver.
+        solver_kwargs :
+            Additional keyword arguments for the solver.
+
+        Raises
+        ------
+        NameError
+            If any of the solver keyword arguments are not valid.
+        """
+        solver_args = inspect.getfullargspec(getattr(jaxopt, solver_name)).args
+        undefined_kwargs = set(solver_kwargs.keys()).difference(solver_args)
+        if undefined_kwargs:
+            raise NameError(
+                f"kwargs {undefined_kwargs} in solver_kwargs not a kwarg for jaxopt.{solver_name}!"
+            )
+
+    @staticmethod
+    def _check_is_callable_from_jax(func: Callable):
+        """
+        Check if the provided function is callable and from the jax namespace.
+
+        Ensures that the given function is not only callable, but also belongs to
+        the `jax` namespace, ensuring compatibility and safety when using jax-based
+        operations.
+
+        Parameters
+        ----------
+        func :
+            The function to check.
+
+        Raises
+        ------
+        TypeError
+            If the provided function is not callable.
+        ValueError
+            If the function does not belong to the `jax` or `neurostatslib.glm` namespaces.
+        """
+        if not callable(func):
+            raise TypeError("The loss function must a Callable!")
+
+        if (not hasattr(func, "__module__")) or (
+            not (
+                func.__module__.startswith("jax")
+                or func.__module__.startswith("neurostatslib.glm")
+            )
+        ):
+            raise ValueError(
+                f"The function {func.__name__} is not from the jax namespace. "
+                "Only functions from the jax namespace are allowed."
+            )
+
+    @abc.abstractmethod
+    def instantiate_solver(
+        self,
+        loss: Callable[
+            [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jnp.ndarray
+        ],
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """Abstract method to instantiate a solver with a given loss function."""
+        pass
+
+    def get_runner(
```

Comment:
looks like this always gets called from within `instantiate_solver`. do we want to just have it live in the superclass version of `instantiate_solver` and have the subclasses call `super().instantiate_solver()` at the bottom?

since this method never gets called by itself and anyone implementing their own solver should call it in `instantiate_solver()`